src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+    //
+    // It's possible for a block to be skipped altogether during serialization,
+    // because it referred to some value which could not be converted into
+    // valid SB3 data. For reporters, this means leaving an empty input; for
+    // stack blocks, it means skipping to the next block in the sibling array
+    // (or leaving an empty connection if there is none). It's up to the caller
+    // to handle serializeBlock returning a null blockId usefully.
+
+    const blockData = newBlockData();
+
+    const { getBroadcastId, getCustomBlockData, parent, siblingBlocks, stage, target } = options;
+
+    let nextBlock: Block;
+    let nextBlockId: sb3.Block["next"];
+    if (siblingBlocks) {
+      const thisIndex = siblingBlocks.indexOf(block) + 1;
+      nextBlock = siblingBlocks.find((x, index) => index === thisIndex);
+    }
+
+    if (nextBlock) {
+      const { blockData: nextBlockData, blockId } = serializeBlock(nextBlock, {
+        stage,
+        target,
+        getBroadcastId,
+        getCustomBlockData,
+        parent: block,
+        siblingBlocks
+      });
+
+      applyBlockData(blockData, nextBlockData);
+      nextBlockId = blockId;
+    }
+
+    const serializeInputsResult = serializeInputs(block, {
+      stage,
+      target,
+
+      getBroadcastId,
+      getCustomBlockData
+    });
+
+    if (!serializeInputsResult) {
+      return { blockData, blockId: nextBlockId };
+    }
+
+    const { inputs, fields, mutation, blockData: inputBlockData } = serializeInputsResult;
+
+    applyBlockData(blockData, inputBlockData);
+
+    const obj: sb3.Block = {
+      opcode: block.opcode,
+
+      parent: parent ? parent.id : null,
+      next: nextBlockId,
+      topLevel: !parent,
+
+      inputs,
+      fields,
+      mutation,
+
+      shadow: false
+    };
+
+    if (obj.topLevel) {
+      obj.x = options.x;
+      obj.y = options.y;
+    }
+
+    const blockId = block.id;
+
+    blockData[blockId] = obj;
+
+    return { blockData, blockId };
+  }
+
+  interface CustomBlockData {
+    args: Array<{
+      default: string;
+      id: string;
+      name: string;
+      type: "boolean" | "numberOrString";
+    }>;
+    warp: boolean;
+  }
+
+  type GetCustomBlockData = (proccode: string) => CustomBlockData;
+
+  type BaseGetBroadcastId = (name: string) => string;
+  interface GetBroadcastId extends BaseGetBroadcastId {
+    initialBroadcastName: string;
+  }
+
+  function collectCustomBlockData(target: Target): GetCustomBlockData {
+    // Parse the scripts in a target, collecting metadata about each custom
+    // block's arguments and other info, and return a function for accessing
+    // the data associated with a particular custom block's proccode.
+    //
+    // It's necesary to collect this data prior to serializing any associated
+    // procedure_call blocks, because they require access to data only found
+    // on the associated procedure_definition. (Specifically, the types of
```

Comment:
```suggestion
    // procedures_call blocks, because they require access to data only found
    // on the associated procedures_definition. (Specifically, the types of
```